src/devices/ShiftRegister/ShiftRegister.cs

```diff
@@ -0,0 +1,260 @@
+            _controller.Write(_sdi, value);
+            // data is written to the storage register on the rising edge of the storage register clock
+            _controller.Write(_clk, 1);
+            // values are reset to low
+            _controller.Write(_sdi, 0);
+            _controller.Write(_clk, 0);
+        }
+
+        /// <summary>
+        /// Shifts a byte -- 8 bits -- to the storage register.
+        /// Assumes register bit length evenly divisible by 8.
+        /// Pushes / overwrites any existing values.
+        /// Latches by default.
+        /// </summary>
+        public void ShiftByte(byte value, bool latch = true)
+        {
+            if (_spiDevice is object)
+            {
+                _spiDevice.WriteByte(value);
+                return;
+            }
+
+            for (int i = 0; i < 8; i++)
+            {
+                // 0b_1000_0000 (same as integer 128) used as input to create mask
+                // determines value of i bit in byte value
+                // logical equivalent of value[i] (which isn't supported for byte type in C#)
+                // starts left-most and ends up right-most
+                int data = (0b_1000_0000 >> i) & value;
+                // writes value to storage register
+                ShiftBit(data);
+            }
+
+            if (latch)
+            {
+                Latch();
+            }
+        }
+
+        /// <summary>
+        /// Latches values in data register to output pi.
+        /// Requires use of GPIO controller.
+        /// </summary>
+        public void Latch()
+        {
+            if (_controller is null || _pinMapping.LE == 0)
+            {
+                throw new ArgumentNullException($"{nameof(Latch)}: GpioController was not provided or {nameof(_pinMapping.LE)} not mapped to pin");
+            }
+
+            // latches value on rising edge of register clock (LE)
+            _controller.Write(_latch, 1);
+            // value reset to low in preparation for next use.
+            _controller.Write(_latch, 0);
+        }
+
+        /// <summary>
+        /// Switch output register to high-impedance state.
+        /// Temporarily disables register outputs, but does not delete values.
+        /// Requires use of GPIO controller.
+        /// </summary>
+        public void OutputDisable()
+        {
+            if (_controller is null || _pinMapping.OE == 0)
+            {
+                throw new ArgumentNullException($"{nameof(OutputDisable)}: {nameof(_pinMapping.OE)} not mapped to non-zero pin value");
+            }
+
+            _controller.Write(_pinMapping.OE, 1);
+        }
+
+        /// <summary>
+        /// Switch output register low-impedance state.
+        /// Enables register outputs with existing values.
+        /// Requires use of GPIO controller.
+        /// </summary>
+        public void OutputEnable()
+        {
+            if (_controller is null || _pinMapping.OE == 0)
```

Comment:
Same comment as somewhere else, maybe check <=0?
And thinking of it, 0 is a valid GPIO. So for all this, should be <0